kv/kvclient/rangefeed: TestRangefeedCatchupStarvation failed

kv/kvclient/rangefeed.TestRangefeedCatchupStarvation [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/19151771?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/19151771?buildTab=artifacts#/) on master @ [af1c7708535272b466c160bbb57abb96efb0b3cd](https://github.com/cockroachdb/cockroach/commits/af1c7708535272b466c160bbb57abb96efb0b3cd):


```
github.com/cockroachdb/cockroach/pkg/rpc.serverStreamInterceptorsChain.run.func1({0x50cdf60?, 0x40054aca88?}, {0x5d89588?, 0x4001319cb0?})
	pkg/rpc/context.go:866 +0x8c
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func2.1({0x5?, 0x5d58040?})
	pkg/rpc/context.go:142 +0x34
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0x4005d542d0, {0x5d58040, 0x4009561720}, {0x5225705, 0x28}, 0x4007df9de8)
	pkg/util/stop/stopper.go:350 +0xec
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func2({0x50cdf60, 0x40054aca88}, {0x5d89588, 0x4001319cb0}, 0x8ff02b0, 0x400c4fbb40)
	pkg/rpc/context.go:141 +0xb4
github.com/cockroachdb/cockroach/pkg/rpc.serverStreamInterceptorsChain.run.func1({0x50cdf60?, 0x40054aca88?}, {0x5d89588?, 0x4001319cb0?})
	pkg/rpc/context.go:866 +0x8c
github.com/cockroachdb/cockroach/pkg/rpc.serverStreamInterceptorsChain.run({0x4008f341c0, 0x4, 0x4}, {0x50cdf60, 0x40054aca88}, {0x5d89588, 0x4001319cb0}, 0x8ff02b0, 0x400bc3c360)
	pkg/rpc/context.go:868 +0x10c
github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.MuxRangeFeed.func3()
	pkg/rpc/context.go:1036 +0xe0
golang.org/x/sync/errgroup.(*Group).Go.func1()
	external/org_golang_x_sync/errgroup/errgroup.go:78 +0x54
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 91190
	external/org_golang_x_sync/errgroup/errgroup.go:75 +0x98

goroutine 90928 [select]:
github.com/cockroachdb/cockroach/pkg/kv/kvserver/closedts/sidetransport.(*Sender).Run.func2({0x5d58040, 0x4003b2c500})
	pkg/kv/kvserver/closedts/sidetransport/sender.go:251 +0x11c
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0x51f41a8, 0x21})
	pkg/util/stop/stopper.go:500 +0x188
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 56442
	pkg/util/stop/stopper.go:489 +0x36c

goroutine 89360 [select, 4 minutes]:
github.com/cockroachdb/cockroach/pkg/sql.(*runnerCoordinator).init.func1({0x0?, 0x5d58040?})
	pkg/sql/distsql_running.go:163 +0xb0
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0x51866a9, 0xe})
	pkg/util/stop/stopper.go:500 +0x188
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 89286
	pkg/util/stop/stopper.go:489 +0x36c

goroutine 89321 [select, 4 minutes]:
github.com/cockroachdb/cockroach/pkg/sql.(*runnerCoordinator).init.func1({0x0?, 0x5d58040?})
	pkg/sql/distsql_running.go:163 +0xb0
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0x51866a9, 0xe})
	pkg/util/stop/stopper.go:500 +0x188
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 89286
	pkg/util/stop/stopper.go:489 +0x36c

goroutine 89369 [select, 4 minutes]:
github.com/cockroachdb/cockroach/pkg/sql.(*runnerCoordinator).init.func1({0x0?, 0x5d58040?})
	pkg/sql/distsql_running.go:163 +0xb0
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0x51866a9, 0xe})
	pkg/util/stop/stopper.go:500 +0x188
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 89286
	pkg/util/stop/stopper.go:489 +0x36c
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangefeedCatchupStarvation.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-48686

```
=== RUN   TestRangefeedCatchupStarvation/feed_type=mux/buffered_sender=false
-- Test timed out at 2025-03-19 21:27:27 UTC --
panic: test timed out after 5m0s
	running tests:
		TestRangefeedCatchupStarvation (4m4s)
		TestRangefeedCatchupStarvation/feed_type=mux/buffered_sender=false (4m4s)

goroutine 103731 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2373 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:215 +0x38

goroutine 1 [chan receive, 4 minutes]:
testing.(*T).Run(0x4000d66ea0, {0x51db5b9?, 0x40018879f8?}, 0x5519fa0)
	GOROOT/src/testing/testing.go:1751 +0x328
testing.runTests.func1(0x4000d66ea0)
	GOROOT/src/testing/testing.go:2168 +0x40
testing.tRunner(0x4000d66ea0, 0x4001887b18)
	GOROOT/src/testing/testing.go:1690 +0xe4
testing.runTests(0x40019f8bd0, {0x90b0ec0, 0x12, 0x12}, {0x4001887b88?, 0x418f18?, 0x9474500?})
	GOROOT/src/testing/testing.go:2166 +0x3ac
testing.(*M).Run(0x40019cefa0)
	GOROOT/src/testing/testing.go:2034 +0x588
github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed_test.TestMain(0x400141ad90?)
	pkg/kv/kvclient/rangefeed/main_test.go:26 +0x9c
main.main()
	bazel-out/aarch64-fastbuild/bin/pkg/kv/kvclient/rangefeed/rangefeed_test_/testmain.go:170 +0x398

goroutine 42 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.flushDaemon()
	pkg/util/log/log_flush.go:123 +0x50
created by github.com/cockroachdb/cockroach/pkg/util/log.init.5 in goroutine 1
	pkg/util/log/log_flush.go:78 +0x24

goroutine 43 [chan receive, 5 minutes]
```

--- a/pkg/kv/kvclient/rangefeed/rangefeed_external_test.go
+++ b/pkg/kv/kvclient/rangefeed/rangefeed_external_test.go
@@ -1923,10 +1923,14 @@ func TestRangefeedCatchupStarvation(t *testing.T) {
 		f, err := rangefeed.NewFactory(s.AppStopper(), db, s.ClusterSettings(), nil)
 		require.NoError(t, err)
 
-		blocked := make(chan struct{})
+		blocked := make(chan struct{}, 1)
 		r1, err := f.RangeFeed(ctx, "consumer-1-rf-1", []roachpb.Span{span}, ts,
 			func(ctx context.Context, value *kvpb.RangeFeedValue) {
-				blocked <- struct{}{}
+				select {
+				case blocked <- struct{}{}:
+				default:
+					// We can hit this case on retries.
+				}
 				<-ctx.Done()
 			},
 			rangefeed.WithConsumerID(1),

